cEP-0017.md

```diff
@@ -0,0 +1,176 @@
+# cib Design and Improvements
+
+| Metadata |                                                     |
+| -------- | --------------------------------------------------- |
+| cEP      | 17                                                  |
+| Version  | 1.0                                                 |
+| Title    | cib Design and Improvements                         |
+| Authors  | Saksham Bansal <mailto:sakshambansal1999@yahoo.com> |
+| Status   | Implementation due                                  |
+| Type     | Process                                             |
+
+## Abstract
+
+This cEP describes the design of the `cib` tool which will allow it to install
+individual bears and their dependencies easily.
+
+The cEP suggests changes to  `Requirement` classes and using conda packages
+for bears such as `DartLintBear`. The cEP describes breaking `DistributionRequirement`
+into different `Requirement` classes for each seperate platform.
+
+## Introduction
+
+The current design of the cib tool allows it to perform basic operations such as
+running pip install on bears and then checking their dependencies. However, the
+design of the `cib` tool needs to be able to handle other special cases such as of
+bears that have different installation instructions for each distribution such as
+`DartLintBear` and bears that have bear dependencies.
+
+Additionally, the tool should also be able to provide verbose output, ask user for
+permission before proceeding and install bears related to a specific language.
+
+## Proposed Change
+
+### Here is the procedure for the installation of the bears
+
+1. Identify the valid bear names that need to be installed as specified by the
+   user on the command line.
+2. Install bear on the valid bear names. We will run commands from `CondaRequirement`
+   to check whether the package is uploaded to `anaconda.org` and then check whether
+   it can be installed from `PyPi` using `PipRequirement`.
+   - This will require to create conda packages for bears that have different
+     installation instructions for each distribution. This is done by making `bld.bat`
+     file that has installation instruction for Windows and similarly a `bld.sh` file
+     that has installation instruction for Linux.
+3. After installing the bear we proceed to installing its requirements. We call
+   methods from the requirement classes to do this.
+   - Add new Requirement classes to support different dependencies from various linters.
+   - The DistributionRequirement class needs to be split into different classes for each
+     package manager and call its method to introduct more consistency.
+4. We then proceed to install bear dependencies such as for `ClangCloneDetectionBear` by
+   running `install_bears` method on each dependency bear.
```

Comment:
`install_bears` is a function, not a method.

also 'on each' implies some OOP voodoo.  'for each' is probably better.